staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -449,33 +455,49 @@ func (c *AvailableConditionController) deleteAPIService(obj interface{}) {
 		}
 	}
 	klog.V(4).Infof("Deleting %q", castObj.Name)
-	c.enqueue(castObj)
+	if castObj.Spec.Service != nil {
+		c.rebuildAPIServiceCache()
+	}
+	c.enqueue(castObj.Name)
 }
 
-// there aren't very many apiservices, just check them all.
-func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []*apiregistration.APIService {
+func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []string {
 	metadata, err := meta.Accessor(obj)
 	if err != nil {
 		utilruntime.HandleError(err)
 		return nil
 	}
+	c.RLock()
+	defer c.RUnlock()
+	if svcMap, ok := c.cache[metadata.GetNamespace()]; ok {
+		if names, exist := svcMap[metadata.GetName()]; exist {
+			return names
+		}
+	}
+	return nil
+}
 
-	var ret []*apiregistration.APIService
+func (c *AvailableConditionController) rebuildAPIServiceCache() {
```

Comment:
this function needs unit tests

Reply:
available_controller_test doesn't exercise this part of code.

Trying to find an existing test which runs thru this code before adding new test.

Reply:
After some search, I haven't found test which triggers the code path touched by this PR.

Reply:
I am adding TestBuildCache to available_controller_test.go (work in progress)

Reply:
Currently aPIServiceLister#List returns nothing in the new test.
Need to figure out how to populate the list.

Reply:
@liggitt 
I have added a basic test.
Can you take a look ?